### PR TITLE
New feat ts plots and tests

### PR DESCRIPTION
**Pull request by KrishnanSG**
#### Describe the changes you've made
I have added the new time series test module. This module will help perform pre modelling statistical tests and other important time series tests.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Have created a new test case in `test_time_series`

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

## Screenshots

 Original           | Updated
 :--------------------: |:--------------------:
 **original screenshot**  | <b>updated screenshot </b> |

### CODE REVIEW

File: pycaret/internal/ts_tests.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    **kwargs
+):
+    coeffs = np.polyfit(range(len(data)), data, trend_order)
+    slope = coeffs[-2]
+    return slope
+
+
+def trend_and_seasonality_test(
+    ts: Union[pd.DataFrame, pd.Series],
+    seasonal_periods=(3, 5, 7, 12, 14, 30, 31),
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:21]: Let's remove the default values from here. This should come from outside depending on the time series.

File: pycaret/internal/ts_tests.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+        # seasonal plot
+        df.plot.bar(ax=ax[1], rot=0)
+        plt.hlines(1, -1, len(df.index), linestyles='--', colors='black', alpha=0.5)
+        ax[1].legend(['Significance level', 'Seasonal value'])
+        ax[1].set_title('Seasonal plot', size=12)
+
+        f.tight_layout()
+
+    if abs(trend) >= 1:
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:68]: Is there any reason to use 0.1 and 1 as the cutoff for the slope to define strong and weak increasnig?

File: pycaret/internal/ts_tests.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+
+def correlation_tests(
+    ts: Union[pd.DataFrame, pd.Series],
+    plots=False,
+    **kwargs
+):
+    from statsmodels.tsa.stattools import acf, pacf
+
+    acf_x, ci = acf(ts, alpha=0.05, nlags=20)
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:129]: You would want to parameterize alpha and nlags.

File: pycaret/internal/ts_tests.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    lower = ci[:, 0] - acf_x
+    upper = ci[:, 1] - acf_x
+    acf_x = np.round(((acf_x >= upper) + (acf_x <= lower)) * acf_x, 2)
+    acf_map = {val: idx for idx, val in enumerate(acf_x)}
+    acf_values = list(filter(lambda x: abs(x) > 0, acf_x))
+    acf_lags = list(map(lambda x: acf_map[x], acf_values))
+
+    pacf_x, ci = pacf(ts, alpha=0.05, nlags=20)
+    lower = ci[:, 0] - pacf_x
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:138]: Since you are comparing pacf to upper and lower ci below, do we need to subtract it here?

```python
pacf_x >= upper) + (pacf_x <= lower)
```

File: pycaret/internal/ts_tests.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    return result
+
+
+def stationarity_tests(
+    ts: Union[pd.DataFrame, pd.Series],
+    cutoff=0.05,
+    **kwargs
+):
+    from statsmodels.tsa.stattools import adfuller
+    dftest = adfuller(ts, autolag='AIC', maxlag=20)
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:171]: we should parameterize this in the main function call

File: pycaret/internal/ts_tests.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    pvalue = dftest[1]
+    result = {
+        'ADF stationarity': 'Difference Stationary' if pvalue < cutoff else 'Not stationary. Try differencing.',
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:174]: I would remove the "try differencing" since it is not needed for all models.

File: pycaret/internal/ts_tests.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    result.update({
+        'KPSS stationarity': "Not stationary. Try removing trend." if statistic > critical_values[
+            "5%"] else 'Trend Stationary',
+        '_KPSS pvalue': p_value
+    })
+
+    if result['KPSS stationarity'] == 'Trend Stationary' and result['ADF stationarity'] == 'Difference Stationary':
+        result['Stationarity'] = 'Stationary'
+    elif result['KPSS stationarity'] != 'Trend Stationary' and result['ADF stationarity'] != 'Difference Stationary':
+        result['Stationarity'] = 'Not stationary. Try GARCH models.'
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:190]: Lets remove "Try GARCH models" for now since we will automate this in the future.

File: pycaret/internal/ts_tests.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    else:
+        result['Stationarity'] = 'Partial stationarity'
+    return result
+
+
+def ts_tests(
```
Comment (ngupta23) [inline pycaret/internal/ts_tests.py:196]: Let's just call it `tests` instead of `ts_tests`

**Review by ngupta23 — COMMENTED** [review]
For later: Could you add Ljung-Box test and also provide an argument to run specific tests?

File: pycaret/internal/pycaret_experiment/time_series_experiment.py [hunk 9]
```diff
@@ -2719,6 +2713,26 @@ def finalize_model(
             estimator=estimator, fit_kwargs=fit_kwargs, model_only=model_only,
         )
 
+    def ts_tests(
```
Comment (TremaMiguel) [inline pycaret/internal/pycaret_experiment/time_series_experiment.py:2716]: I suggest hear to change the name to `ts_stat_tests` or something that makes reference to calculating an statistical test, otherwise the name is not clear about what type of time series tests is running.

File: pycaret/internal/ts_tests.py (hunk 8 above)
Comment (TremaMiguel) [reply pycaret/internal/ts_tests.py:196]: @ngupta23 can we add to the name something related to statisticall, like `stat_tests` ?

File: pycaret/internal/ts_tests.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    This method performs all statistical tests for the given time series dataset. This is further help in modelling
+    and most modelling inference
+
+    Returns
+    -------
+    Dataframe with the results for the tests
+
+    """
+    test_results = {
+        'Temporal': trend_and_seasonality_test(ts, **kwargs),
```
Comment (TremaMiguel) [inline pycaret/internal/ts_tests.py:211]: If `ts_tests` is the main interface to the user, that is, `trend_and_seasonality_tests`, `correlation_tests` are not directly called by the user, then I suggest to use reference here private methods 

```python
def _trend_and_seasonality_tests():
     pass
```

and then in the main interface 

```python
 test_results = {
        'Temporal': _trend_and_seasonality_test(ts, **kwargs),
```

File: pycaret/internal/ts_tests.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        'Correlation': correlation_tests(ts, **kwargs),
+        'Statistics': statistic_tests(ts, **kwargs),
+        'Stationarity': stationarity_tests(ts, **kwargs)
+    }
+    reform = {(outerKey, innerKey): values for outerKey, innerDict in test_results.items() for innerKey, values in
+              innerDict.items() if (show_hidden_keys or innerKey[0] != '_')}
```
Comment (TremaMiguel) [inline pycaret/internal/ts_tests.py:217]: will `innerKey` always have elements ?

File: pycaret/tests/test_time_series.py [hunk 12]
```diff
@@ -611,3 +609,32 @@ def test_tune_model_raises(load_data):
...
+
+    exp = TimeSeriesExperiment()
+    data = load_data  # _check_data_for_prophet(name, load_data)
+
+    exp.setup(
+        data=data,
+        verbose=False,
+    )
+    result = exp.ts_tests(show_hidden_keys=True)
+    # print(result)
```
Comment (TremaMiguel) [inline pycaret/tests/test_time_series.py:625]: we can also test here if DataFrame is not empty

File: pycaret/tests/test_time_series.py [hunk 13]
```diff
@@ -611,3 +609,32 @@ def test_tune_model_raises(load_data):
...
+        ('Correlation', 'Highest PACF lag'), ('Correlation', '_ACF values'),
+        ('Correlation', '_PACF values'), ('Statistics', 'Mean'), ('Statistics', 'Median'),
+        ('Statistics', 'Standard Deviation'), ('Statistics', 'Variance'),
+        ('Statistics', 'Kurtosis'), ('Statistics', 'Skewness'),
+        ('Statistics', 'Count of distinct values'), ('Statistics', '_Data distribution'),
+        ('Statistics', 'Length'), ('Statistics', 'Shapiro normality test'),
+        ('Statistics', '_Normality pvalue'), ('Stationarity', 'ADF stationarity'),
+        ('Stationarity', '_ADF pvalue'), ('Stationarity', 'KPSS stationarity'),
+        ('Stationarity', '_KPSS pvalue'), ('Stationarity', 'Stationarity')
+    ]
```
Comment (TremaMiguel) [inline pycaret/tests/test_time_series.py:640]: what happens if the values from the result are all the same, can we test the values from the ouput of `exp.ts_tests` ?

Comment (TremaMiguel) [inline pycaret/tests/test_time_series.py:640]: @KrishnanSG added some comments, if you want to check them

**Review by TremaMiguel — CHANGES_REQUESTED** [review]

File: pycaret/internal/ts_tests.py (hunk 8 above)
Comment (ngupta23) [reply pycaret/internal/ts_tests.py:196]: It might be better to have

```python
def tests(type="stats")
```

What do you think?

**Comment by ngupta23** [comment]
@KrishnanSG @TremaMiguel, @moezali1 

Here are my thoughts on how testing and plotting should work. We should maintain consistency with other `pycaret` modules as much as possible. Let me know what you think about this approach.

## Overview

1. Create internal private modular functions 
2. These can be combined at a higher level and exposed to the end users of the package

## Internal functions

### Tests & EDA

```python
eda_(data, eda="summary_stats", *kwargs)
test_(data, test="trend", *kwargs)
test_(data, test="seasonality", *kwargs)
test_(data, test="shapiro", *kwargs)
test_(data, test="stationarity", *kwargs)
    test_(data, test="adf", *kwargs)
    test_(data, test="kpss", *kwargs)
test_(data, test="ljung-box", *kwargs)

# Each Test should returns a DataFrame with respective values indexed appropriately
# This will allow multiple test results to be combined together at a higher level     
```

### Plots

```python
plot_(data, plot="acf", *kwargs)
plot_(data, plot="pacf", *kwargs)
plot_(data, plot="residuals", *kwargs)
```

## Higher Level Integration (for end users)

```python
# Should follow style of plot_model from regression and other existing modules
# e.g. plot_model(lr, plot = 'residual')

test_model(model=None,      test="seasonality")  # Applied to original data. Calls internal test_() / eda_() appropriately 
test_model(model=lr_cds_dt, test="seasonality")  # Applied to model residuals. Calls internal test_() / eda_() appropriately 
test_model(model=None)                           # By default, it can run all tests (on original data) and return the DataFrame of combined results
test_model(model=lr_cds_dt)                      # By default, it can run all tests (on model residuals) and return the DataFrame of combined results

plot_model(model=None,      plot="acf")          # Applied to original data. Calls internal plot_() appropriately
plot_model(model=my_model,  plot="acf")          # Applied to model residuals. Calls internal plot_() appropriately
plot_model(model=my_model,  plot="acf")          # Applied to model residuals. Calls internal plot_() appropriately

plot_model(model=my_model)                       # By default plots predictions and prediction interval of the model along with original time series

plot_model(model=None,      plot="data")         # Plots the original data
plot_model(model=None,      plot="splits-tt")    # Plots the splits (train-test) that will be used
plot_model(model=None,      plot="splits-cv")    # Plots the splits (cv splits inside the train splits) that will be used

# https://github.com/pycaret/pycaret/issues/1405
plot_model(data=results_df, plot="cv") 
# Plots metrics, distributions for results and actual time series predictions for a single model 
# results_df will be returned by create_model

# https://github.com/pycaret/pycaret/issues/1405
plot_model(data=results_df, plot="compare") 
# Plots metrics, distributions for results and actual time series predictions from multiple models 
# results_df will be returned by compare_models
```

**Comment by ngupta23** [comment]
Closed through #1495 and #1544
